File: Cpp_src/medium/add_two_numbers_II_445.cc
#include <iostream>
#include <stack>
using namespace std;

// Definition for singly-linked list.
struct ListNode
{
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

// Use stack
// Time Complexity: O(MAX(m,n))
// Space Complexity: O(MAX(m,n))
// Runtime: 24 ms, faster than 75.67%
// Memory Usage: 13 MB, less than 51.85%
class Solution
{
public:
    ListNode *addTwoNumbers(ListNode *l1, ListNode *l2)
    {
        ListNode *result = new ListNode(0);
        ListNode *head = result;
        stack<int> stack1;
        stack<int> stack2;

        while (l1)
        {
            stack1.push(l1->val);
            l1 = l1->next;
        }

        while (l2)
        {
            stack2.push(l2->val);
            l2 = l2->next;
        }

        int carry = 0;
        while (!stack1.empty() || !stack2.empty() || carry != 0)
        {
            int num1 = 0;
            int num2 = 0;
            if (!stack1.empty())
            {
                num1 = stack1.top();
                stack1.pop();
            }

            if (!stack2.empty())
            {
                num2 = stack2.top();
                stack2.pop();
            }
            int sum = num1 + num2 + carry;
            ListNode *tmp = new ListNode(sum % 10);
            tmp->next = head->next;
            head->next = tmp;

            carry = sum / 10;
        }

        return result->next;
    }
};

File: Golang_src/medium/addTwoNumbersII_445.go
package medium

// Use stack:
// Time Complexity: O(max(m,n))
// Space Complexity:O(m+n)
// Runtime: 2 ms, faster than 97.73%
// Memory Usage: 5.4 MB, less than 44.32%
func addTwoNumbers2(l1 *ListNode, l2 *ListNode) *ListNode {
	stack1, stack2 := []int{}, []int{}
	cur1, cur2 := l1, l2
	for cur1 != nil {
		stack1 = append(stack1, cur1.Val)
		cur1 = cur1.Next
	}
	for cur2 != nil {
		stack2 = append(stack2, cur2.Val)
		cur2 = cur2.Next
	}

	carry := 0
	dummy := &ListNode{Val: -1}
	for len(stack1) > 0 || len(stack2) > 0 || carry != 0 {
		tmp1, tmp2 := 0, 0
		if len(stack1) > 0 {
			tmp1 = stack1[len(stack1)-1]
			stack1 = stack1[:len(stack1)-1]
		}
		if len(stack2) > 0 {
			tmp2 = stack2[len(stack2)-1]
			stack2 = stack2[:len(stack2)-1]
		}
		node := &ListNode{Val: (tmp1 + tmp2 + carry) % 10}
		carry = (tmp1 + tmp2 + carry) / 10
		node.Next = dummy.Next
		dummy.Next = node
	}

	return dummy.Next
}
